### PR TITLE
Add new `Style/MagicCommentFormat` cop

### PR DESCRIPTION
**Pull request by mattbearman**
Magic comments in Ruby are underscore/hyphen agnostic, this can lead to codebases having inconsistent case style in the frozen string literal comment (`# frozen_string_literal: true` and `# frozen-string-literal: true`)

This cop allows the enforcement of either snake case (underscore separated) or kebab case (hyphen separated).

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] ~~Commit message starts with `[Fix #issue-number]` (if the related issue exists).~~
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop-hq/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
I think it'd better is this was a cop covering all such pragmas. One cop per pragma seems like a massive overkill to me.

**Comment by mattbearman** [comment]
That's a great point, I'll see about extending this to all magic comments 👍

**Comment by bbatsov** [comment]
@mattbearman ping :-)

**Comment by dvandersluis** [comment]
@mattbearman are you still interested in completing this?

**Comment by dvandersluis** [comment]
@bbatsov I have updated this PR to complete it.

The cop is now called `Style/MagicComments`, and it enforces consistency on all magic comments, not just `frozen-string-literal`. I also added the ability to keep capitalization consistent, for both the comment directives and values independently.

**Comment by dvandersluis** [comment]
Some tests are failing randomly and I haven't been able to repro yet. Looking into it.

**Comment by dvandersluis** [comment]
Ah, figured it out, `filter_map` isn't available in all rubies :grin: Fixed it now.

File: config/default.yml [hunk 1]
```diff
@@ -4015,6 +4015,26 @@ Style/LineEndConcatenation:
...
+  SupportedStyles:
+    # `snake` will enforce the magic comment is written
+    # in snake case (words separated by underscores).
+    # Eg: froze_string_literal: true
+    - snake
+    # `kebab` will enforce the magic comment is written
+    # in kebab case (words separated by hyphens).
+    # Eg: froze-string-literal: true
+    - kebab
+  DirectiveCapitalization: lower
```
Comment (bbatsov) [inline config/default.yml:4032]: Seems to me this can simply be a boolean.

Comment (dvandersluis) [reply config/default.yml:4032]: There are three options - `upper` (all uppercase), `lower` (all lowercase) and nil (ignore capitalization)

**Comment by koic** [comment]
`Style/MagicComments` may be abstract. A somewhat concreted name like `Style/MagicCommentFormat`,  `Style/MagicCommentShape`, or more better name looks user friendly.

File: config/default.yml [hunk 2]
```diff
@@ -4015,6 +4015,26 @@ Style/LineEndConcatenation:
...
+Style/MagicComments:
+  Description: 'Use a consistent style for magic comments.'
+  Enabled: true
+  VersionAdded: '<<next>>'
+  EnforcedStyle: snake
+  SupportedStyles:
+    # `snake` will enforce the magic comment is written
+    # in snake case (words separated by underscores).
+    # Eg: froze_string_literal: true
+    - snake
```
Comment (koic) [inline config/default.yml:4027]: `snake_case` is consistent with `Naming/MethodName`, `Naming/VariableNumber` option names.

```suggestion
    - snake_case
```

- https://docs.rubocop.org/rubocop/1.34/cops_naming.html#namingmethodname
- https://docs.rubocop.org/rubocop/1.34/cops_naming.html#namingvariablenumber

File: config/default.yml [hunk 3]
```diff
@@ -4015,6 +4015,26 @@ Style/LineEndConcatenation:
   VersionAdded: '0.18'
   VersionChanged: '0.64'
 
+Style/MagicComments:
+  Description: 'Use a consistent style for magic comments.'
+  Enabled: true
```
Comment (koic) [inline config/default.yml:4020]:
```suggestion
  Enabled: pending
```

File: config/default.yml (hunk 2 above)
Comment (dvandersluis) [reply config/default.yml:4027]: Works for me, I'll fix them up.

**Comment by dvandersluis** [comment]
> `Style/MagicComments` may be abstract. A somewhat concreted name like `Style/MagicCommentFormat`, `Style/MagicCommentShape`, or more better name looks user friendly.

Sure, `MagicCommentFormat` works for me.

File: config/default.yml (hunk 1 above)
Comment (bbatsov) [reply config/default.yml:4032]: Yeah, I saw this, but was on the fence for whether it makes sense for someone to disable this. I've yet to see someone not writing lowercase magic comments, but I guess they exist. :D

File: config/default.yml [hunk 4]
```diff
@@ -4015,6 +4015,26 @@ Style/LineEndConcatenation:
...
+    # Eg: froze_string_literal: true
+    - snake_case
+    # `kebab` will enforce the magic comment is written
+    # in kebab case (words separated by hyphens).
+    # Eg: froze-string-literal: true
+    - kebab_case
+  DirectiveCapitalization: lower
+  ValueCapitalization: ~
+  SupportedCapitalizations:
+    - lower
```
Comment (bbatsov) [inline config/default.yml:4035]: I think I like `DirectiveCase` a bit better as `capitalization` and `lowercase` don't mix very well IMO. :-) Perhaps spelling out fully `lowercase` and `uppercase` would be best.

Comment (dvandersluis) [reply config/default.yml:4035]: I had `DirectiveCase` first but I found it confusing with snake case and kebab case. I'll change it to the full `lowercase` and `uppercase`, good suggestion!

File: config/default.yml (hunk 1 above)
Comment (dvandersluis) [reply config/default.yml:4032]: I figured that since it's easy, having options is better :smile:

File: config/default.yml (hunk 4 above)
Comment (dvandersluis) [reply config/default.yml]: Updated!

**Pull request merged**